File: platform/packages_mjs/fetch_undici_gemini_openai_protocols/src/streaming.ts
/**
 * Streaming Format Translation
 *
 * Handles translation between OpenAI SSE streaming format and Gemini streaming format.
 * Provides async iterators for real-time translation of streaming responses.
 */

import { Readable } from 'stream'
import { logger, iterSSE, iterLines } from 'fetch-undici'
import type {
  OpenAIStreamChunk,
  OpenAIDelta,
  OpenAIStreamChoice,
  GeminiGenerateResponse,
  GeminiCandidate,
  GeminiContent,
  TranslationResult,
} from './types.js'
import { mapGeminiFinishReason } from './response.js'
import { isGeminiFunctionCallPart, generateToolCallId } from './tools.js'
import { mapGeminiRoleToOpenAI } from './messages.js'

const log = logger.create('fetch-undici-gemini-openai-protocols:streaming', import.meta.url)

// =============================================================================
// OpenAI Streaming Types (extended for streaming)
// =============================================================================

/**
 * State for accumulating streamed content
 */
interface StreamAccumulator {
  id: string
  model: string
  content: string
  toolCalls: Map<number, {
    id: string
    name: string
    arguments: string
  }>
  role?: string
  finishReason: string | null
}

// =============================================================================
// Gemini → OpenAI Streaming Translation
// =============================================================================

/**
 * Convert a single Gemini streaming chunk to OpenAI format
 */
export function geminiChunkToOpenAI(
  chunk: GeminiGenerateResponse,
  options: {
    id?: string
    model?: string
    isFirst?: boolean
    idGenerator?: () => string
  } = {}
): TranslationResult<OpenAIStreamChunk> {
  const warnings: string[] = []
  const {
    id = `chatcmpl-${Date.now()}`,
    model = 'gemini-model',
    isFirst = false,
    idGenerator = generateToolCallId,
  } = options

  const choices: OpenAIStreamChoice[] = []

  for (const candidate of chunk.candidates || []) {
    const delta: OpenAIDelta = {}

    // On first chunk, include the role
    if (isFirst) {
      delta.role = mapGeminiRoleToOpenAI(candidate.content?.role || 'model')
    }

    // Extract text content
    if (candidate.content?.parts) {
      const textParts = candidate.content.parts.filter(
        (p): p is { text: string } => 'text' in p
      )
      if (textParts.length > 0) {
        delta.content = textParts.map(p => p.text).join('')
      }

      // Extract function calls
      const functionCallParts = candidate.content.parts.filter(isGeminiFunctionCallPart)
      if (functionCallParts.length > 0) {
        delta.tool_calls = functionCallParts.map((part, index) => ({
          index,
          id: idGenerator(),
          type: 'function' as const,
          function: {
            name: part.functionCall.name,
            arguments: JSON.stringify(part.functionCall.args),
          },
        }))
      }
    }

    choices.push({
      index: candidate.index ?? 0,
      delta,
      finish_reason: candidate.finishReason
        ? mapGeminiFinishReason(candidate.finishReason)
        : null,
    })
  }

  return {
    data: {
      id,
      object: 'chat.completion.chunk',
      created: Math.floor(Date.now() / 1000),
      model,
      choices,
    },
    warnings,
  }
}

/**
 * Create an async iterator that converts Gemini NDJSON stream to OpenAI SSE format
 */
export async function* geminiStreamToOpenAI(
  stream: Readable,
  options: {
    model?: string
    idGenerator?: () => string
  } = {}
): AsyncGenerator<string> {
  log.debug('Starting Gemini to OpenAI stream translation')

  const { model = 'gemini-model', idGenerator = generateToolCallId } = options
  const id = `chatcmpl-${Date.now()}-${Math.random().toString(36).substring(2, 9)}`

  let isFirst = true
  let chunkCount = 0

  for await (const line of iterLines(stream, 'utf-8')) {
    const trimmed = line.trim()
    if (!trimmed) continue

    try {
      const geminiChunk = JSON.parse(trimmed) as GeminiGenerateResponse
      const { data: openAIChunk } = geminiChunkToOpenAI(geminiChunk, {
        id,
        model,
        isFirst,
        idGenerator,
      })

      chunkCount++
      isFirst = false

      // Yield in SSE format
      yield `data: ${JSON.stringify(openAIChunk)}\n\n`
    } catch (err) {
      log.warn('Failed to parse Gemini streaming chunk', {
        error: (err as Error).message,
        line: trimmed.substring(0, 100),
      })
    }
  }

  // Send [DONE] marker
  yield 'data: [DONE]\n\n'

  log.debug('Gemini to OpenAI stream translation complete', { chunkCount })
}

// =============================================================================
// OpenAI → Gemini Streaming Translation
// =============================================================================

/**
 * Convert OpenAI SSE chunk to Gemini format
 */
export function openAIChunkToGemini(
  chunk: OpenAIStreamChunk,
  accumulator: StreamAccumulator
): TranslationResult<GeminiGenerateResponse | null> {
  const warnings: string[] = []

  const candidates: GeminiCandidate[] = []

  for (const choice of chunk.choices) {
    const { delta, finish_reason } = choice

    // Update accumulator
    if (delta.role) {
      accumulator.role = delta.role
    }

    if (delta.content) {
      accumulator.content += delta.content
    }

    // Handle tool calls (accumulated)
    if (delta.tool_calls) {
      for (const tc of delta.tool_calls) {
        const existing = accumulator.toolCalls.get(tc.index)
        if (existing) {
          if (tc.function?.arguments) {
            existing.arguments += tc.function.arguments
          }
        } else {
          accumulator.toolCalls.set(tc.index, {
            id: tc.id || generateToolCallId(),
            name: tc.function?.name || '',
            arguments: tc.function?.arguments || '',
          })
        }
      }
    }

    if (finish_reason) {
      accumulator.finishReason = finish_reason
    }

    // Build Gemini content from current delta
    const parts: GeminiContent['parts'] = []

    if (delta.content) {
      parts.push({ text: delta.content })
    }

    // Only emit function calls when complete (on finish)
    if (finish_reason && accumulator.toolCalls.size > 0) {
      for (const [, tc] of accumulator.toolCalls) {
        let args: Record<string, unknown>
        try {
          args = JSON.parse(tc.arguments)
        } catch {
          args = {}
        }
        parts.push({
          functionCall: {
            name: tc.name,
            args,
          },
        })
      }
    }

    if (parts.length > 0 || finish_reason) {
      candidates.push({
        content: {
          role: 'model',
          parts,
        },
        finishReason: finish_reason
          ? (['stop', 'tool_calls', 'function_call'].includes(finish_reason) ? 'STOP' : 'MAX_TOKENS')
          : undefined,
        index: choice.index,
      })
    }
  }

  // Only return a response if we have candidates
  if (candidates.length === 0) {
    return { data: null, warnings }
  }

  return {
    data: { candidates },
    warnings,
  }
}

/**
 * Create an async iterator that converts OpenAI SSE stream to Gemini NDJSON format
 */
export async function* openAIStreamToGemini(
  stream: Readable,
  _options: {
    model?: string
  } = {}
): AsyncGenerator<string> {
  log.debug('Starting OpenAI to Gemini stream translation')

  const accumulator: StreamAccumulator = {
    id: '',
    model: '',
    content: '',
    toolCalls: new Map(),
    finishReason: null,
  }

  let chunkCount = 0

  for await (const event of iterSSE(stream)) {
    // Skip empty data
    if (!event.data || event.data === '[DONE]') {
      continue
    }

    try {
      const openAIChunk = JSON.parse(event.data) as OpenAIStreamChunk

      // Update accumulator metadata
      if (!accumulator.id && openAIChunk.id) {
        accumulator.id = openAIChunk.id
      }
      if (!accumulator.model && openAIChunk.model) {
        accumulator.model = openAIChunk.model
      }

      const { data: geminiChunk } = openAIChunkToGemini(openAIChunk, accumulator)

      if (geminiChunk) {
        chunkCount++
        yield JSON.stringify(geminiChunk) + '\n'
      }
    } catch (err) {
      log.warn('Failed to parse OpenAI streaming chunk', {
        error: (err as Error).message,
      })
    }
  }

  log.debug('OpenAI to Gemini stream translation complete', { chunkCount })
}

// =============================================================================
// Stream Aggregation Utilities
// =============================================================================

/**
 * Aggregate OpenAI streaming chunks into a complete response
 */
export async function aggregateOpenAIStream(
  stream: Readable
): Promise<TranslationResult<OpenAIStreamChunk>> {
  log.debug('Aggregating OpenAI stream')

  const warnings: string[] = []
  let id = ''
  let model = ''
  let created = 0
  const content: string[] = []
  const toolCalls = new Map<number, {
    id: string
    type: 'function'
    function: { name: string; arguments: string }
  }>()
  let finishReason: string | null = null
  let role: string | undefined

  for await (const event of iterSSE(stream)) {
    if (!event.data || event.data === '[DONE]') continue

    try {
      const chunk = JSON.parse(event.data) as OpenAIStreamChunk

      // Capture metadata
      if (!id) id = chunk.id
      if (!model) model = chunk.model
      if (!created) created = chunk.created

      for (const choice of chunk.choices) {
        if (choice.delta.role && !role) {
          role = choice.delta.role
        }
        if (choice.delta.content) {
          content.push(choice.delta.content)
        }
        if (choice.delta.tool_calls) {
          for (const tc of choice.delta.tool_calls) {
            const existing = toolCalls.get(tc.index)
            if (existing) {
              if (tc.function?.arguments) {
                existing.function.arguments += tc.function.arguments
              }
            } else {
              toolCalls.set(tc.index, {
                id: tc.id || generateToolCallId(),
                type: 'function',
                function: {
                  name: tc.function?.name || '',
                  arguments: tc.function?.arguments || '',
                },
              })
            }
          }
        }
        if (choice.finish_reason) {
          finishReason = choice.finish_reason
        }
      }
    } catch (err) {
      warnings.push(`Failed to parse chunk: ${(err as Error).message}`)
    }
  }

  // Build final aggregated chunk
  const delta: OpenAIDelta = {}
  if (role) delta.role = role as OpenAIDelta['role']
  if (content.length > 0) delta.content = content.join('')
  if (toolCalls.size > 0) {
    delta.tool_calls = Array.from(toolCalls.entries()).map(([index, tc]) => ({
      index,
      id: tc.id,
      type: tc.type,
      function: tc.function,
    }))
  }

  return {
    data: {
      id,
      object: 'chat.completion.chunk',
      created,
      model,
      choices: [{
        index: 0,
        delta,
        finish_reason: finishReason,
      }],
    },
    warnings,
  }
}

/**
 * Aggregate Gemini NDJSON stream into a complete response
 */
export async function aggregateGeminiStream(
  stream: Readable
): Promise<TranslationResult<GeminiGenerateResponse>> {
  log.debug('Aggregating Gemini stream')

  const warnings: string[] = []
  const candidates = new Map<number, {
    content: GeminiContent
    finishReason?: GeminiCandidate['finishReason']
  }>()
  let usageMetadata: GeminiGenerateResponse['usageMetadata'] | undefined

  for await (const line of iterLines(stream, 'utf-8')) {
    const trimmed = line.trim()
    if (!trimmed) continue

    try {
      const chunk = JSON.parse(trimmed) as GeminiGenerateResponse

      // Aggregate candidates
      for (const candidate of chunk.candidates || []) {
        const index = candidate.index ?? 0
        const existing = candidates.get(index)

        if (existing) {
          // Merge parts
          if (candidate.content?.parts) {
            existing.content.parts.push(...candidate.content.parts)
          }
          if (candidate.finishReason) {
            existing.finishReason = candidate.finishReason
          }
        } else {
          candidates.set(index, {
            content: candidate.content || { role: 'model', parts: [] },
            finishReason: candidate.finishReason,
          })
        }
      }

      // Capture usage (usually in last chunk)
      if (chunk.usageMetadata) {
        usageMetadata = chunk.usageMetadata
      }
    } catch (err) {
      warnings.push(`Failed to parse chunk: ${(err as Error).message}`)
    }
  }

  // Build final response
  const finalCandidates: GeminiCandidate[] = Array.from(candidates.entries()).map(
    ([index, data]) => ({
      content: data.content,
      finishReason: data.finishReason,
      index,
    })
  )

  return {
    data: {
      candidates: finalCandidates,
      usageMetadata,
    },
    warnings,
  }
}

File: platform/packages_py/fetch_httpx_gemini_openai_protocols/fetch_httpx_gemini_openai_protocols/streaming.py
"""
Streaming Format Translation

Handles translation between OpenAI SSE streaming format and Gemini streaming format.
Provides async iterators for real-time translation of streaming responses.
"""

from __future__ import annotations

import json
import random
import time
from collections.abc import AsyncIterator, Callable
from dataclasses import dataclass, field
from typing import Any

from fetch_httpx import create_logger

from .messages import map_gemini_role_to_openai
from .response import map_gemini_finish_reason
from .tools import generate_tool_call_id, is_gemini_function_call_part
from .types import (
    GeminiCandidate,
    GeminiGenerateResponse,
    OpenAIDelta,
    OpenAIStreamChoice,
    OpenAIStreamChunk,
    TranslationResult,
)

logger = create_logger("fetch_httpx_gemini_openai_protocols.streaming", __file__)


# =============================================================================
# Stream Accumulator
# =============================================================================


@dataclass
class StreamAccumulator:
    """State for accumulating streamed content."""

    id: str = ""
    model: str = ""
    content: str = ""
    tool_calls: dict[int, dict[str, Any]] = field(default_factory=dict)
    role: str | None = None
    finish_reason: str | None = None


# =============================================================================
# Gemini → OpenAI Streaming Translation
# =============================================================================


def gemini_chunk_to_openai(
    chunk: GeminiGenerateResponse,
    id: str | None = None,
    model: str = "gemini-model",
    is_first: bool = False,
    id_generator: Callable[[], str] | None = None,
) -> TranslationResult[OpenAIStreamChunk]:
    """Convert a single Gemini streaming chunk to OpenAI format."""
    warnings: list[str] = []

    if id is None:
        id = f"chatcmpl-{int(time.time())}"

    if id_generator is None:
        id_generator = generate_tool_call_id

    choices: list[OpenAIStreamChoice] = []

    for candidate in chunk.get("candidates", []):
        delta: OpenAIDelta = {}

        content = candidate.get("content", {})

        # On first chunk, include the role
        if is_first:
            delta["role"] = map_gemini_role_to_openai(content.get("role", "model"))

        parts = content.get("parts", [])

        # Extract text content
        text_parts = [p for p in parts if "text" in p]
        if text_parts:
            delta["content"] = "".join(p["text"] for p in text_parts)

        # Extract function calls
        function_call_parts = [p for p in parts if is_gemini_function_call_part(p)]
        if function_call_parts:
            delta["tool_calls"] = [
                {
                    "index": idx,
                    "id": id_generator(),
                    "type": "function",
                    "function": {
                        "name": part["functionCall"]["name"],
                        "arguments": json.dumps(part["functionCall"]["args"]),
                    },
                }
                for idx, part in enumerate(function_call_parts)
            ]

        finish_reason = candidate.get("finishReason")
        choices.append({
            "index": candidate.get("index", 0),
            "delta": delta,
            "finish_reason": map_gemini_finish_reason(finish_reason) if finish_reason else None,
        })

    return TranslationResult(
        data={
            "id": id,
            "object": "chat.completion.chunk",
            "created": int(time.time()),
            "model": model,
            "choices": choices,
        },
        warnings=warnings,
    )


async def gemini_stream_to_openai(
    stream: AsyncIterator[str],
    model: str = "gemini-model",
    id_generator: Callable[[], str] | None = None,
) -> AsyncIterator[str]:
    """
    Async iterator that converts Gemini NDJSON stream to OpenAI SSE format.

    Args:
        stream: Async iterator yielding Gemini NDJSON lines
        model: Model name to use in response
        id_generator: Optional function to generate tool call IDs

    Yields:
        OpenAI SSE formatted strings (data: {...}\n\n)
    """
    logger.debug("Starting Gemini to OpenAI stream translation")

    random_suffix = "".join(random.choices("abcdefghijklmnopqrstuvwxyz0123456789", k=7))
    stream_id = f"chatcmpl-{int(time.time())}-{random_suffix}"

    is_first = True
    chunk_count = 0

    async for line in stream:
        trimmed = line.strip()
        if not trimmed:
            continue

        try:
            gemini_chunk = json.loads(trimmed)
            result = gemini_chunk_to_openai(
                gemini_chunk,
                id=stream_id,
                model=model,
                is_first=is_first,
                id_generator=id_generator,
            )

            chunk_count += 1
            is_first = False

            # Yield in SSE format
            yield f"data: {json.dumps(result.data)}\n\n"

        except json.JSONDecodeError as e:
            logger.warn(
                "Failed to parse Gemini streaming chunk",
                context={"error": str(e), "line": trimmed[:100]},
            )

    # Send [DONE] marker
    yield "data: [DONE]\n\n"

    logger.debug("Gemini to OpenAI stream translation complete", context={"chunk_count": chunk_count})


# =============================================================================
# OpenAI → Gemini Streaming Translation
# =============================================================================


def openai_chunk_to_gemini(
    chunk: OpenAIStreamChunk,
    accumulator: StreamAccumulator,
) -> TranslationResult[GeminiGenerateResponse | None]:
    """Convert OpenAI SSE chunk to Gemini format."""
    warnings: list[str] = []

    candidates: list[GeminiCandidate] = []

    for choice in chunk.get("choices", []):
        delta = choice.get("delta", {})
        finish_reason = choice.get("finish_reason")

        # Update accumulator
        if "role" in delta:
            accumulator.role = delta["role"]

        if "content" in delta:
            accumulator.content += delta["content"]

        # Handle tool calls (accumulated)
        tool_calls = delta.get("tool_calls", [])
        for tc in tool_calls:
            index = tc.get("index", 0)
            existing = accumulator.tool_calls.get(index)
            if existing:
                func = tc.get("function", {})
                if "arguments" in func:
                    existing["arguments"] += func["arguments"]
            else:
                func = tc.get("function", {})
                accumulator.tool_calls[index] = {
                    "id": tc.get("id", generate_tool_call_id()),
                    "name": func.get("name", ""),
                    "arguments": func.get("arguments", ""),
                }

        if finish_reason:
            accumulator.finish_reason = finish_reason

        # Build Gemini content from current delta
        parts: list[dict[str, Any]] = []

        if "content" in delta:
            parts.append({"text": delta["content"]})

        # Only emit function calls when complete (on finish)
        if finish_reason and accumulator.tool_calls:
            for tc in accumulator.tool_calls.values():
                try:
                    args = json.loads(tc["arguments"])
                except json.JSONDecodeError:
                    args = {}
                parts.append({
                    "functionCall": {
                        "name": tc["name"],
                        "args": args,
                    }
                })

        if parts or finish_reason:
            gemini_finish = None
            if finish_reason:
                if finish_reason in ("stop", "tool_calls", "function_call"):
                    gemini_finish = "STOP"
                else:
                    gemini_finish = "MAX_TOKENS"

            candidates.append({
                "content": {
                    "role": "model",
                    "parts": parts,
                },
                "finishReason": gemini_finish,
                "index": choice.get("index", 0),
            })

    if not candidates:
        return TranslationResult(data=None, warnings=warnings)

    return TranslationResult(
        data={"candidates": candidates},
        warnings=warnings,
    )


async def openai_stream_to_gemini(
    stream: AsyncIterator[str],
    model: str = "gemini-model",
) -> AsyncIterator[str]:
    """
    Async iterator that converts OpenAI SSE stream to Gemini NDJSON format.

    Args:
        stream: Async iterator yielding OpenAI SSE lines
        model: Model name (unused, for API consistency)

    Yields:
        Gemini NDJSON formatted strings
    """
    logger.debug("Starting OpenAI to Gemini stream translation")

    accumulator = StreamAccumulator()
    chunk_count = 0

    # Parse SSE format
    async for line in stream:
        line = line.strip()

        if line.startswith("data: "):
            data = line[6:]  # Remove "data: " prefix

            if data == "[DONE]":
                continue

            try:
                openai_chunk = json.loads(data)

                # Update accumulator metadata
                if not accumulator.id and openai_chunk.get("id"):
                    accumulator.id = openai_chunk["id"]
                if not accumulator.model and openai_chunk.get("model"):
                    accumulator.model = openai_chunk["model"]

                result = openai_chunk_to_gemini(openai_chunk, accumulator)

                if result.data:
                    chunk_count += 1
                    yield json.dumps(result.data) + "\n"

            except json.JSONDecodeError as e:
                logger.warn(
                    "Failed to parse OpenAI streaming chunk",
                    context={"error": str(e)},
                )

    logger.debug("OpenAI to Gemini stream translation complete", context={"chunk_count": chunk_count})


# =============================================================================
# Stream Aggregation Utilities
# =============================================================================


async def aggregate_openai_stream(
    stream: AsyncIterator[str],
) -> TranslationResult[OpenAIStreamChunk]:
    """Aggregate OpenAI streaming chunks into a complete response."""
    logger.debug("Aggregating OpenAI stream")

    warnings: list[str] = []
    id = ""
    model = ""
    created = 0
    content: list[str] = []
    tool_calls: dict[int, dict[str, Any]] = {}
    finish_reason: str | None = None
    role: str | None = None

    async for line in stream:
        line = line.strip()

        if not line.startswith("data: "):
            continue

        data = line[6:]
        if data == "[DONE]":
            continue

        try:
            chunk = json.loads(data)

            # Capture metadata
            if not id:
                id = chunk.get("id", "")
            if not model:
                model = chunk.get("model", "")
            if not created:
                created = chunk.get("created", 0)

            for choice in chunk.get("choices", []):
                delta = choice.get("delta", {})

                if "role" in delta and not role:
                    role = delta["role"]

                if "content" in delta:
                    content.append(delta["content"])

                for tc in delta.get("tool_calls", []):
                    index = tc.get("index", 0)
                    existing = tool_calls.get(index)
                    if existing:
                        func = tc.get("function", {})
                        if "arguments" in func:
                            existing["function"]["arguments"] += func["arguments"]
                    else:
                        tool_calls[index] = {
                            "id": tc.get("id", generate_tool_call_id()),
                            "type": "function",
                            "function": {
                                "name": tc.get("function", {}).get("name", ""),
                                "arguments": tc.get("function", {}).get("arguments", ""),
                            },
                        }

                if choice.get("finish_reason"):
                    finish_reason = choice["finish_reason"]

        except json.JSONDecodeError as e:
            warnings.append(f"Failed to parse chunk: {e}")

    # Build final aggregated chunk
    delta: OpenAIDelta = {}
    if role:
        delta["role"] = role
    if content:
        delta["content"] = "".join(content)
    if tool_calls:
        delta["tool_calls"] = [
            tool_calls[idx] for idx in sorted(tool_calls.keys())
        ]

    return TranslationResult(
        data={
            "id": id,
            "object": "chat.completion.chunk",
            "created": created,
            "model": model,
            "choices": [{
                "index": 0,
                "delta": delta,
                "finish_reason": finish_reason,
            }],
        },
        warnings=warnings,
    )


async def aggregate_gemini_stream(
    stream: AsyncIterator[str],
) -> TranslationResult[GeminiGenerateResponse]:
    """Aggregate Gemini NDJSON stream into a complete response."""
    logger.debug("Aggregating Gemini stream")

    warnings: list[str] = []
    candidates: dict[int, dict[str, Any]] = {}
    usage_metadata: dict[str, int] | None = None

    async for line in stream:
        trimmed = line.strip()
        if not trimmed:
            continue

        try:
            chunk = json.loads(trimmed)

            # Aggregate candidates
            for candidate in chunk.get("candidates", []):
                index = candidate.get("index", 0)
                existing = candidates.get(index)

                if existing:
                    # Merge parts
                    content = candidate.get("content", {})
                    if content.get("parts"):
                        existing["content"]["parts"].extend(content["parts"])
                    if candidate.get("finishReason"):
                        existing["finishReason"] = candidate["finishReason"]
                else:
                    candidates[index] = {
                        "content": candidate.get("content", {"role": "model", "parts": []}),
                        "finishReason": candidate.get("finishReason"),
                        "index": index,
                    }

            # Capture usage (usually in last chunk)
            if chunk.get("usageMetadata"):
                usage_metadata = chunk["usageMetadata"]

        except json.JSONDecodeError as e:
            warnings.append(f"Failed to parse chunk: {e}")

    # Build final response
    final_candidates = [
        candidates[idx] for idx in sorted(candidates.keys())
    ]

    result: GeminiGenerateResponse = {
        "candidates": final_candidates,
    }

    if usage_metadata:
        result["usageMetadata"] = usage_metadata

    return TranslationResult(data=result, warnings=warnings)
